{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymol import cmd \n",
    "import numpy as np\n",
    "import sys\n",
    "import os\n",
    "\n",
    "ref_structure = \"../homologues_plastic/SP3.pdb\"\n",
    "infilenames = [\"../homologues_plastic/SP1.pdb\",\"../homologues_plastic/SP2.pdb\",\"../homologues_plastic/SP3.pdb\",\"../homologues_plastic/A0A1B6ZCQ7.cif\"]\n",
    "\n",
    "#@title Download options\n",
    "download_pymol = False #@param {type: \"boolean\"}\n",
    "#@markdown `download_pymol` allows you to download a pymol file of your aligned structures. \n",
    "outfile_name = \"tmp/outfile\" #@param {type:\"string\"}\n",
    "#@markdown Choose name of the pymol file.\n",
    "download_alignment = False #@param {type: \"boolean\"}\n",
    "#@markdown `download_alignment` allows you to download a alignment file of your aligned structures in sequence format.\n",
    "alignment_file_name = \"tmp/alignment.aln\" #@param {type:\"string\"}\n",
    "#@markdown Choose name of the alignment file.\n",
    "\n",
    "# removing spaces from outfile and add .pse\n",
    "outfile_name = \"_\".join(outfile_name.split(\" \"))+\".pse\"\n",
    "\n",
    "#@title Color options\n",
    "color_mode = \"similarity\" #@param [\"similarity\", \"hotspot\", \"none\"] {type:\"string\"}\n",
    "#@markdown `color_mode` specify which way the structures should be colored.\n",
    "\n",
    "#@title SIMalign options\n",
    "max_iterations = 3 #@param {type:\"integer\"}\n",
    "#@markdown `max_iterations` is the maximum number of alignments. A high number can lead to slow runtime.\n",
    "min_aligned_aa = 100 #@param {type:\"integer\"}\n",
    "#@markdown `min_aligned_aa` is how many amino acid that minimum should be used for alignment. A low number can lead to overfitting.\n",
    "max_dist = 6 #@param {type:\"integer\"}\n",
    "#@markdown `max_dist` is the maximum length between to amino acids before it is considered as a gab in the alignment. A too low number can lead to false gabs and a too high number can lead to false positive.\n",
    "# remove_chain_duplicate = True #@param {type:\"boolean\"}\n",
    "# For now it only takes chain A\n",
    "#  #@markdown If `remove_chain_duplicate` is true then is chain duplicates removed from the structure.\n",
    "\n",
    "#@title Foldseek options\n",
    "#@markdown Foldseek doesn't work yet!\n",
    "foldseek = False #@param {type:\"boolean\"}\n",
    "foldseek_database = \"Alphafold/UniProt\" #@param [\"Alphafold/UniProt\",\"Alphafold/UniProt50-minimal\",\"Alphafold/UniProt50\",\"Alphafold/Proteome\",\"Alphafold/Swiss-Prot\",\"ESMAtlas30\",\"PDB\",\"Thermophilic_DB\"] {type:\"string\"}\n",
    "foldseek_variable_tresshold = \"evalue\" #@param [\"evalue\",\"pident\",\"fident\",\"nident\",\"alnlen\",\"bits\",\"mismatch\",\"qcov\",\"tcov\",\"lddt\",\"qtmscore\",\"ttmscore\",\"alntmscore\",\"rmsd\",\"prob\"]\n",
    "foldseek_value_tresshold = 0.0001  #@param {type:\"number\"}\n",
    "foldseek_search_against = \"ref_structure\" #@param [\"ref_structure\",\"all_structures\"]\n",
    "\n",
    "#@title Hotspot finding options\n",
    "find_hotspots = True #@param {type: \"boolean\"}\n",
    "#@markdown If `find_hotspots` is true, then the program will find amino acid in the structure that can be mutated to potentially alter the stability of the protein.\n",
    "#hotspot_min_size = 2 #@param {type: \"integer\"}\n",
    "#For now we only finds single mutations\n",
    "\n",
    "#@title Run prediction\n",
    "import SIMalign\n",
    "\n",
    "\n",
    "if len(infilenames) < 3 and foldseek == False:\n",
    "    print(\"ERROR: Import at least 3 files or activate foldseek.\")\n",
    "    sys.exit(1)\n",
    "\n",
    "if foldseek:\n",
    "    import foldseek_search\n",
    "    foldseek_search.run(foldseek_database,foldseek_variable_tresshold,foldseek_value_tresshold,foldseek_search_against,ref_structure,infilenames)\n",
    "len_ref_structure, score_list, structure_list, core_selection = SIMalign.run(ref_structure, infilenames, max_iterations, min_aligned_aa, max_dist, alignment_file_name)\n",
    "\n",
    "if find_hotspots:\n",
    "    import hotspot_finder\n",
    "    hotspot_list, exposed_list = hotspot_finder.run(structure_list,alignment_file_name)\n",
    "    hotspot_finder.print_hotspot(hotspot_list,structure_list)\n",
    "else:\n",
    "    hotspot_list, exposed_list = None, None\n",
    "\n",
    "\n",
    "if color_mode == \"hotspot\" and find_hotspots == False:\n",
    "    print(\"ERROR: unable to color hotspot without finding them!\")\n",
    "else:\n",
    "    import colors\n",
    "    colors.run(color_mode,hotspot_list,score_list,structure_list,core_selection,exposed_list)\n",
    "\n",
    "from pymol import cmd\n",
    "cmd.save(outfile_name)\n",
    "if download_pymol:\n",
    "    files.download(outfile_name)\n",
    "else:\n",
    "    print(\"Done\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
