{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading structures to pymol...\n",
      "Running SIMalign...\n",
      "\tsuperimposing SP1 towards SP3\n",
      "\tsuperimposing SP2 towards SP3\n",
      "\tsuperimposing A0A1B6ZCQ7 towards SP3\n",
      "\tIteration 1\n",
      "\tstructure\tRMSD\tresidues aligned\n",
      "\tSP1\t0.542\t331\n",
      "\tSP2\t0.379\t382\n",
      "\tA0A1B6ZCQ7\t0.708\t381\n",
      "\tsuperimposing SP1 towards SP3\n",
      "\tsuperimposing SP2 towards SP3\n",
      "\tsuperimposing A0A1B6ZCQ7 towards SP3\n",
      "\tIteration 2\n",
      "\tstructure\tRMSD\tresidues aligned\n",
      "\tSP1\t0.486\t186\n",
      "\tSP2\t0.349\t203\n",
      "\tA0A1B6ZCQ7\t0.604\t189\n",
      "\tCompleted 2 iteration(s) of superimposion.\n",
      "Finding hotspots...\n",
      "\tPrinting possible single mutations in SP3\n",
      "\t\tCYS24 -> THR as structure: SP1\n",
      "\t\tCYS24 -> THR as structure: A0A1B6ZCQ7\n",
      "\t\tILE75 -> VAL as structure: A0A1B6ZCQ7\n",
      "\t\tSER78 -> ALA as structure: SP1\n",
      "\t\tILE79 -> PHE as structure: SP2\n",
      "\t\tILE79 -> PHE as structure: A0A1B6ZCQ7\n",
      "\t\tSER102 -> ALA as structure: SP1\n",
      "\t\tTHR119 -> SER as structure: SP2\n",
      "\t\tILE121 -> VAL as structure: SP1\n",
      "\t\tSER130 -> ALA as structure: A0A1B6ZCQ7\n",
      "\t\tALA148 -> PRO as structure: SP1\n",
      "\t\tALA148 -> PRO as structure: A0A1B6ZCQ7\n",
      "\t\tALA155 -> SER as structure: SP1\n",
      "\t\tLEU166 -> VAL as structure: A0A1B6ZCQ7\n",
      "\t\tTHR190 -> PRO as structure: SP1\n",
      "\t\tMET218 -> LEU as structure: A0A1B6ZCQ7\n",
      "\t\tLEU248 -> SER as structure: A0A1B6ZCQ7\n",
      "\t\tCYS276 -> LEU as structure: SP1\n",
      "\t\tCYS351 -> ALA as structure: SP1\n",
      "\t\tGLY355 -> SER as structure: A0A1B6ZCQ7\n",
      "\t\tPHE381 -> TYR as structure: SP1\n",
      "\t\tALA386 -> GLY as structure: SP2\n",
      "\t\tILE388 -> ALA as structure: SP1\n",
      "\t\tILE388 -> ALA as structure: A0A1B6ZCQ7\n",
      "\t\tSER389 -> GLY as structure: SP1\n",
      "\t\tSER389 -> GLY as structure: A0A1B6ZCQ7\n",
      "\t\tLEU399 -> VAL as structure: SP1\n",
      "\t\tLEU399 -> THR as structure: A0A1B6ZCQ7\n",
      "\t\tSER400 -> VAL as structure: A0A1B6ZCQ7\n",
      "\t\tMET401 -> VAL as structure: SP1\n",
      "\t\tMET401 -> LEU as structure: A0A1B6ZCQ7\n",
      "\t\tGLY413 -> ASN as structure: SP2\n",
      "\t\tILE429 -> GLY as structure: A0A1B6ZCQ7\n",
      "\tPrinting possible single mutations in SP1\n",
      "\t\tTHR25 -> CYS as structure: SP3\n",
      "\t\tTHR25 -> CYS as structure: SP2\n",
      "\t\tALA79 -> SER as structure: SP3\n",
      "\t\tALA79 -> SER as structure: SP2\n",
      "\t\tALA79 -> SER as structure: A0A1B6ZCQ7\n",
      "\t\tPHE80 -> ILE as structure: SP3\n",
      "\t\tVAL82 -> ILE as structure: A0A1B6ZCQ7\n",
      "\t\tALA103 -> SER as structure: SP2\n",
      "\t\tALA103 -> SER as structure: A0A1B6ZCQ7\n",
      "\t\tILE115 -> VAL as structure: SP3\n",
      "\t\tILE115 -> VAL as structure: SP2\n",
      "\t\tILE115 -> VAL as structure: A0A1B6ZCQ7\n",
      "\t\tVAL122 -> ILE as structure: SP3\n",
      "\t\tVAL122 -> ILE as structure: SP2\n",
      "\t\tVAL122 -> ILE as structure: A0A1B6ZCQ7\n",
      "\t\tSER131 -> ALA as structure: A0A1B6ZCQ7\n",
      "\t\tTHR148 -> SER as structure: SP3\n",
      "\t\tTHR148 -> SER as structure: SP2\n",
      "\t\tPRO149 -> ALA as structure: SP3\n",
      "\t\tPRO149 -> ALA as structure: SP2\n",
      "\t\tSER156 -> ALA as structure: SP3\n",
      "\t\tILE179 -> VAL as structure: SP3\n",
      "\t\tILE179 -> VAL as structure: SP2\n",
      "\t\tILE179 -> VAL as structure: A0A1B6ZCQ7\n",
      "\t\tPRO191 -> THR as structure: SP2\n",
      "\t\tMET219 -> LEU as structure: A0A1B6ZCQ7\n",
      "\t\tMET229 -> ALA as structure: SP3\n",
      "\t\tMET229 -> ALA as structure: SP2\n",
      "\t\tLEU230 -> PHE as structure: A0A1B6ZCQ7\n",
      "\t\tCYS263 -> VAL as structure: SP3\n",
      "\t\tALA270 -> VAL as structure: A0A1B6ZCQ7\n",
      "\t\tVAL347 -> ILE as structure: SP3\n",
      "\t\tILE348 -> LEU as structure: A0A1B6ZCQ7\n",
      "\t\tPRO351 -> VAL as structure: A0A1B6ZCQ7\n",
      "\t\tALA352 -> LEU as structure: SP3\n",
      "\t\tALA353 -> GLY as structure: SP3\n",
      "\t\tALA353 -> GLY as structure: SP2\n",
      "\t\tALA380 -> SER as structure: SP3\n",
      "\t\tALA380 -> GLY as structure: A0A1B6ZCQ7\n",
      "\t\tHIS381 -> GLN as structure: SP3\n",
      "\t\tALA385 -> ILE as structure: SP3\n",
      "\t\tALA385 -> ILE as structure: SP2\n",
      "\t\tVAL396 -> LEU as structure: SP3\n",
      "\t\tVAL396 -> LEU as structure: SP2\n",
      "\t\tVAL396 -> THR as structure: A0A1B6ZCQ7\n",
      "\t\tALA426 -> GLY as structure: SP3\n",
      "\t\tALA426 -> GLY as structure: SP2\n",
      "\tPrinting possible single mutations in SP2\n",
      "\t\tCYS24 -> THR as structure: SP1\n",
      "\t\tCYS24 -> THR as structure: A0A1B6ZCQ7\n",
      "\t\tILE75 -> VAL as structure: A0A1B6ZCQ7\n",
      "\t\tSER78 -> ALA as structure: SP1\n",
      "\t\tPHE79 -> ILE as structure: SP3\n",
      "\t\tSER102 -> ALA as structure: SP1\n",
      "\t\tSER119 -> THR as structure: SP3\n",
      "\t\tSER119 -> THR as structure: SP1\n",
      "\t\tILE121 -> VAL as structure: SP1\n",
      "\t\tALA148 -> PRO as structure: SP1\n",
      "\t\tALA148 -> PRO as structure: A0A1B6ZCQ7\n",
      "\t\tALA155 -> SER as structure: SP1\n",
      "\t\tTHR190 -> PRO as structure: SP1\n",
      "\t\tMET218 -> LEU as structure: A0A1B6ZCQ7\n",
      "\t\tTHR265 -> VAL as structure: SP3\n",
      "\t\tCYS276 -> LEU as structure: SP1\n",
      "\t\tCYS351 -> ALA as structure: SP1\n",
      "\t\tLEU354 -> SER as structure: A0A1B6ZCQ7\n",
      "\t\tILE388 -> ALA as structure: SP1\n",
      "\t\tILE388 -> ALA as structure: A0A1B6ZCQ7\n",
      "\t\tSER389 -> GLY as structure: SP1\n",
      "\t\tSER389 -> GLY as structure: A0A1B6ZCQ7\n",
      "\t\tLEU399 -> VAL as structure: SP1\n",
      "\t\tLEU399 -> THR as structure: A0A1B6ZCQ7\n",
      "\t\tSER400 -> VAL as structure: A0A1B6ZCQ7\n",
      "\t\tMET401 -> VAL as structure: SP1\n",
      "\t\tASN413 -> GLY as structure: SP3\n",
      "\t\tASN413 -> GLY as structure: SP1\n",
      "\t\tASN413 -> GLY as structure: A0A1B6ZCQ7\n",
      "\t\tILE429 -> GLY as structure: A0A1B6ZCQ7\n",
      "\tPrinting possible single mutations in A0A1B6ZCQ7\n",
      "\t\tTHR30 -> CYS as structure: SP3\n",
      "\t\tTHR30 -> CYS as structure: SP2\n",
      "\t\tSER84 -> ALA as structure: SP1\n",
      "\t\tPHE85 -> ILE as structure: SP3\n",
      "\t\tSER108 -> ALA as structure: SP1\n",
      "\t\tVAL120 -> ILE as structure: SP1\n",
      "\t\tILE121 -> PHE as structure: SP3\n",
      "\t\tILE121 -> PHE as structure: SP2\n",
      "\t\tILE127 -> VAL as structure: SP1\n",
      "\t\tALA136 -> SER as structure: SP3\n",
      "\t\tALA136 -> SER as structure: SP1\n",
      "\t\tALA136 -> SER as structure: SP2\n",
      "\t\tPRO154 -> ALA as structure: SP3\n",
      "\t\tPRO154 -> ALA as structure: SP2\n",
      "\t\tVAL172 -> LEU as structure: SP3\n",
      "\t\tVAL172 -> LEU as structure: SP2\n",
      "\t\tTYR188 -> PHE as structure: SP1\n",
      "\t\tILE200 -> ALA as structure: SP1\n",
      "\t\tILE200 -> ALA as structure: SP2\n",
      "\t\tLEU235 -> ALA as structure: SP3\n",
      "\t\tLEU235 -> ALA as structure: SP2\n",
      "\t\tVAL259 -> LEU as structure: SP3\n",
      "\t\tMET271 -> VAL as structure: SP3\n",
      "\t\tILE275 -> VAL as structure: SP3\n",
      "\t\tILE275 -> VAL as structure: SP2\n",
      "\t\tVAL278 -> ALA as structure: SP3\n",
      "\t\tVAL278 -> ALA as structure: SP1\n",
      "\t\tTYR354 -> ALA as structure: SP3\n",
      "\t\tTYR354 -> ALA as structure: SP1\n",
      "\t\tMET355 -> ILE as structure: SP3\n",
      "\t\tMET355 -> VAL as structure: SP1\n",
      "\t\tSER361 -> GLY as structure: SP3\n",
      "\t\tSER361 -> GLY as structure: SP2\n",
      "\t\tALA391 -> GLY as structure: SP2\n",
      "\t\tALA393 -> ILE as structure: SP3\n",
      "\t\tALA393 -> ILE as structure: SP2\n",
      "\t\tLEU395 -> VAL as structure: SP1\n",
      "\t\tVAL396 -> ALA as structure: SP3\n",
      "\t\tVAL396 -> ALA as structure: SP1\n",
      "\t\tVAL396 -> ALA as structure: SP2\n",
      "\t\tTHR404 -> LEU as structure: SP3\n",
      "\t\tTHR404 -> LEU as structure: SP2\n",
      "\t\tVAL405 -> SER as structure: SP3\n",
      "\t\tVAL405 -> SER as structure: SP2\n",
      "\t\tILE408 -> VAL as structure: SP3\n",
      "\t\tTHR430 -> ALA as structure: SP3\n",
      "\t\tTHR430 -> ALA as structure: SP2\n",
      "\t\tALA434 -> GLY as structure: SP3\n",
      "\t\tALA434 -> GLY as structure: SP2\n",
      "Coloring and selecting in pymol...\n",
      "\tColoring by similarity:\n",
      "\t\tColoring SP3\n",
      "\t\tColoring SP1\n",
      "\t\tColoring SP2\n",
      "\t\tColoring A0A1B6ZCQ7\n",
      "Done\n"
     ]
    }
   ],
   "source": [
    "from pymol import cmd \n",
    "import numpy as np\n",
    "import sys\n",
    "import os\n",
    "\n",
    "ref_structure = \"../homologues_plastic/SP3.pdb\"\n",
    "infilenames = [\"../homologues_plastic/SP1.pdb\",\"../homologues_plastic/SP2.pdb\",\"../homologues_plastic/SP3.pdb\",\"../homologues_plastic/A0A1B6ZCQ7.cif\"]\n",
    "\n",
    "#@title Download options\n",
    "download_pymol = False #@param {type: \"boolean\"}\n",
    "#@markdown `download_pymol` allows you to download a pymol file of your aligned structures. \n",
    "outfile_name = \"tmp/outfile\" #@param {type:\"string\"}\n",
    "#@markdown Choose name of the pymol file.\n",
    "download_alignment = False #@param {type: \"boolean\"}\n",
    "#@markdown `download_alignment` allows you to download a alignment file of your aligned structures in sequence format.\n",
    "alignment_file_name = \"tmp/alignment.aln\" #@param {type:\"string\"}\n",
    "#@markdown Choose name of the alignment file.\n",
    "\n",
    "# removing spaces from outfile and add .pse\n",
    "outfile_name = \"_\".join(outfile_name.split(\" \"))+\".pse\"\n",
    "\n",
    "#@title Color options\n",
    "color_mode = \"similarity\" #@param [\"similarity\", \"hotspot\", \"none\"] {type:\"string\"}\n",
    "#@markdown `color_mode` specify which way the structures should be colored.\n",
    "\n",
    "#@title SIMalign options\n",
    "max_iterations = 2 #@param {type:\"integer\"}\n",
    "#@markdown `max_iterations` is the maximum number of alignments. A high number can lead to slow runtime.\n",
    "min_aligned_aa = 100 #@param {type:\"integer\"}\n",
    "#@markdown `min_aligned_aa` is how many amino acid that minimum should be used for alignment. A low number can lead to overfitting.\n",
    "max_dist = 6 #@param {type:\"integer\"}\n",
    "#@markdown `max_dist` is the maximum length between to amino acids before it is considered as a gab in the alignment. A too low number can lead to false gabs and a too high number can lead to false positive.\n",
    "# remove_chain_duplicate = True #@param {type:\"boolean\"}\n",
    "# For now it only takes chain A\n",
    "#  #@markdown If `remove_chain_duplicate` is true then is chain duplicates removed from the structure.\n",
    "\n",
    "#@title Foldseek options\n",
    "#@markdown Foldseek doesn't work yet!\n",
    "foldseek = False #@param {type:\"boolean\"}\n",
    "foldseek_database = \"Alphafold/UniProt\" #@param [\"Alphafold/UniProt\",\"Alphafold/UniProt50-minimal\",\"Alphafold/UniProt50\",\"Alphafold/Proteome\",\"Alphafold/Swiss-Prot\",\"ESMAtlas30\",\"PDB\",\"Thermophilic_DB\"] {type:\"string\"}\n",
    "foldseek_variable_tresshold = \"evalue\" #@param [\"evalue\",\"pident\",\"fident\",\"nident\",\"alnlen\",\"bits\",\"mismatch\",\"qcov\",\"tcov\",\"lddt\",\"qtmscore\",\"ttmscore\",\"alntmscore\",\"rmsd\",\"prob\"]\n",
    "foldseek_value_tresshold = 0.0001  #@param {type:\"number\"}\n",
    "foldseek_search_against = \"ref_structure\" #@param [\"ref_structure\",\"all_structures\"]\n",
    "\n",
    "#@title Hotspot finding options\n",
    "find_hotspots = True #@param {type: \"boolean\"}\n",
    "#@markdown If `find_hotspots` is true, then the program will find amino acid in the structure that can be mutated to potentially alter the stability of the protein.\n",
    "#hotspot_min_size = 2 #@param {type: \"integer\"}\n",
    "#For now we only finds single mutations\n",
    "\n",
    "#@title Visualization options\n",
    "color_mode = \"similarity\" #@param [\"similarity\", \"hotspot\", \"none\"] {type:\"string\"}\n",
    "#@markdown  - `color_mode` specify which way the structures should be colored.\n",
    "structure_format = \"spheres-sticks\" #@param [\"spheres-sticks\",\"cartoon\",\"spheres\",\"sticks\"] {type:\"string\"}\n",
    "#@markdown   - `structure_format` specify how the structure should be showed in pymol.\n",
    "show_in_pymol = \"only_not_conserved\" #@param [\"only_not_conserved_core\",\"only_core\",\"entire_chain_A\",\"everything\"] {type:\"string\"}\n",
    "#@markdown   - `show_in_pymol` specify what part of the structures that will be shown in pymol.\n",
    "color_by_element = True #@param {type: \"boolean\"}\n",
    "#@markdown   - If `color_by_element` is ON then atom will be colored by element in pymol.\n",
    "\n",
    "\n",
    "#@title Run prediction\n",
    "import SIMalign\n",
    "\n",
    "\n",
    "if len(infilenames) < 3 and foldseek == False:\n",
    "    print(\"ERROR: Import at least 3 files or activate foldseek.\")\n",
    "    sys.exit(1)\n",
    "\n",
    "if foldseek:\n",
    "    import foldseek_search\n",
    "    foldseek_search.run(foldseek_database,foldseek_variable_tresshold,foldseek_value_tresshold,foldseek_search_against,ref_structure,infilenames)\n",
    "len_ref_structure, score_list, structure_list, core_selection = SIMalign.run(ref_structure, infilenames, max_iterations, min_aligned_aa, max_dist, alignment_file_name)\n",
    "\n",
    "if find_hotspots:\n",
    "    import hotspot_finder\n",
    "    hotspot_list, exposed_list = hotspot_finder.run(structure_list,alignment_file_name)\n",
    "    hotspot_finder.print_hotspot(hotspot_list,structure_list)\n",
    "else:\n",
    "    hotspot_list, exposed_list = None, None\n",
    "\n",
    "\n",
    "if color_mode == \"hotspot\" and find_hotspots == False:\n",
    "    print(\"ERROR: unable to color hotspot without finding them!\")\n",
    "else:\n",
    "    import visualization\n",
    "    visualization.run(color_mode,hotspot_list,score_list,structure_list,core_selection,exposed_list,structure_format,show_in_pymol,color_by_element)\n",
    "\n",
    "from pymol import cmd\n",
    "cmd.save(outfile_name)\n",
    "if download_pymol:\n",
    "    files.download(outfile_name)\n",
    "else:\n",
    "    print(\"Done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "426\n"
     ]
    }
   ],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
